pkg/minikube/tunnel/tunnel.go

```diff
@@ -0,0 +1,160 @@
+	"k8s.io/minikube/pkg/minikube/config"
+)
+
+type tunnel interface {
+	cleanup() *Status
+	updateTunnelStatus() *Status
+}
+
+func errorTunnelAlreadyExists(id *ID) error {
+	return fmt.Errorf("there is already a running tunnel for this machine: %s", id)
+}
+
+func newTunnel(machineName string,
+	machineAPI libmachine.API,
+	configLoader config.Loader,
+	v1Core v1.CoreV1Interface, registry *persistentRegistry, router router) (*minikubeTunnel, error) {
+	clusterInspector := &clusterInspector{
+		machineName:  machineName,
+		machineAPI:   machineAPI,
+		configLoader: configLoader,
+	}
+	state, route, err := clusterInspector.getStateAndRoute()
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to determine cluster info: %s", err)
+	}
+	id := ID{
+		Route:       route,
+		MachineName: machineName,
+		Pid:         getPid(),
+	}
+	runningTunnel, err := registry.IsAlreadyDefinedAndRunning(&id)
+	if err != nil {
+		return nil, fmt.Errorf("unable to check tunnel registry for conflict: %s", err)
+	}
+	if runningTunnel != nil {
+		return nil, fmt.Errorf("another tunnel is already running, shut it down first: %s", runningTunnel)
+	}
+
+	return &minikubeTunnel{
+		clusterInspector:     clusterInspector,
+		router:               router,
+		registry:             registry,
+		loadBalancerEmulator: newLoadBalancerEmulator(v1Core),
+		status: &Status{
+			TunnelID:      id,
+			MinikubeState: state,
+		},
+		reporter: &simpleReporter{
+			out: os.Stdout,
+		},
+	}, nil
+
+}
+
+type minikubeTunnel struct {
+	//collaborators
+	clusterInspector     *clusterInspector
+	router               router
+	loadBalancerEmulator loadBalancerEmulator
+	reporter             reporter
+	registry             *persistentRegistry
+
+	status *Status
+}
+
+func (t *minikubeTunnel) cleanup() *Status {
+	glog.V(3).Infof("cleaning up %s", t.status.TunnelID.Route)
+	e := t.router.Cleanup(t.status.TunnelID.Route)
+	if e != nil {
+		t.status.RouteError = errors.Errorf("error cleaning up route: %v", e)
+		glog.V(3).Infof(t.status.RouteError.Error())
+	} else {
+		e = t.registry.Remove(t.status.TunnelID.Route)
+		if e != nil {
+			glog.V(3).Infof("error removing route from registry: %v", e)
+		}
+	}
+	if t.status.MinikubeState == Running {
+		t.status.PatchedServices, t.status.LoadBalancerEmulatorError = t.loadBalancerEmulator.Cleanup()
+	}
+	return t.status
+}
+
+func (t *minikubeTunnel) updateTunnelStatus() *Status {
+	glog.V(3).Info("updating tunnel status...")
+	t.status.MinikubeState, _, t.status.MinikubeError = t.clusterInspector.getStateAndHost()
+	//TODO(balintp): clean this up to be more self contained
+	defer t.clusterInspector.machineAPI.Close()
+	if t.status.MinikubeState == Running {
+		glog.V(3).Infof("minikube is running, trying to add Route %s", t.status.TunnelID.Route)
+
+		exists, conflict, _, err := t.router.Inspect(t.status.TunnelID.Route)
+		if err != nil {
+			t.status.RouteError = fmt.Errorf("error checking for route state: %s", err)
```

Comment:
Can this call return now? That would help to de-indent the non-error flow.

Same elsewhere in this function. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow